### PR TITLE
chore: enable prettier for .md/.mdx files - close #844

### LINKED ISSUE #844: [discussion] use `prettier` for `.md/.mdx` also
### Subject of the feature

As title, I'm not quite sure why we are not using `prettier` for `mdx`.

### Problem

With [remark-preset-prettier](https://github.com/JounQin/remark-preset-prettier) we can disable all rules conflict with `prettier`.

### Expected behavior

Enable `prettier` everywhere.

### PR DESCRIPTION
**Pull request by JounQin**
@mdx-js/maintainers Please help to review whether should we do this.

### CODE REVIEW

File: docs/404.mdx [hunk 1]
```diff
@@ -1,3 +1,3 @@
 # 404
 
-Page not found.  [Return Home](/)
```
Comment (wooorm) [inline docs/404.mdx]: This PR seems to break all paragraphs?

Comment (JounQin) [reply docs/404.mdx]: It seems weird to me, I'll try to figure out.

Comment (JounQin) [reply docs/404.mdx]:
```
retext-sentence-spacing:double-space - Expected `2` spaces between sentences, not `1`
```
🤣 I should upgrade `remark-preset-prettier` first.

Comment (wooorm) [reply docs/404.mdx]: Huh, why are you turning spaces and quotes off? They’re there for a reason.
If prettier is removing all spaces, then that’s a problem with prettier?

Comment (wooorm) [reply docs/404.mdx]: If you want 1 space between sentences, you can change the setting instead of turning it off

Comment (JounQin) [reply docs/404.mdx]: > If prettier is removing all spaces, then that’s a problem with prettier?

It seems comprehensive effect of `eslint`, `prettier` and `remark-lint` together.

Comment (wooorm) [reply docs/404.mdx]: > If we're using prettier for all files, they should be handlered by prettier only.

There’s both ESLint and Prettier, right? So why ditch remark?

Comment (JounQin) [reply docs/404.mdx]: @wooorm It seems I can not override/disable `retext-sentence-spacing` at all by following:

```json
[
  "retext-sentence-spacing",
  {
    "preferred": 1
  }
]
```

Or

```json
[
  "retext-sentence-spacing",
  false
]
```

Any advice again? 🤣

Comment (JounQin) [reply docs/404.mdx]: > > If we're using prettier for all files, they should be handlered by prettier only.
> 
> There’s both ESLint and Prettier, right? So why ditch remark?

https://github.com/mdx-js/eslint-mdx/blob/master/packages/eslint-plugin-mdx/src/rules/remark.ts#L38-L87

`remark` is under the hook of ESLint rule `mdx/remark`.

Comment (wooorm) [reply docs/404.mdx]: Where are you doing this?

Comment (wooorm) [reply docs/404.mdx]: > remark is under the hook of ESLint rule mdx/remark.
 
Right, I know. What I mean is maybe best explained like this: this is a project that uses ESLint. Now, you are adding prettier. Any rule that prettier does differently, you completely remove from ESLint. This is what you are doing by turning of remark-lint rules. You are removing ESLint rules.

I think it’s better to have both ESLint and Prettier. Which can be done by changing the ESLint rules (a.k.a., changing the remark lint rules)

Comment (JounQin) [reply docs/404.mdx]: > Where are you doing this?

`remarkConfig` field in `package.json`: https://github.com/mdx-js/mdx/blob/chore/prettier/package.json#L128 and then `eslint-plugin-mdx` will read it, so I should check `eslint-plugin-mdx`.

Comment (JounQin) [reply docs/404.mdx]: @wooorm `eslint-plugin-prettier` is `prettier` for ESLint. `remark` and `prettier` rules could conflict, so `prettier` recommends disable all rules which may conflict with `prettier` like https://github.com/prettier/eslint-config-prettier.

Comment (wooorm) [reply docs/404.mdx]: Could you provide more information, so I can better help you? What *exactly* is it you are doing.

I don’t understand why you mention `eslint-plugin-mdx`, we’re talking about markdown/mdx here (prose), not jsx/mdx.

Comment (JounQin) [reply docs/404.mdx]: `mdx` files are now formatted by ESLint with `eslint-plugin-prettier` and `eslint-plugin-mdx` via `retext/remark` related plugins under the hook.

Comment (wooorm) [reply docs/404.mdx]: How interesting. Well, this is getting really complicated. I suggest removing *everything* that is now in `remarkConfig`, and replacing it with the plugins you want to use instead.

Why use lots of plugins, then turn most of them off, and add more? Wouldn’t it be clearer to just add the things you want?

Comment (JounQin) [reply docs/404.mdx]: > Why use lots of plugins, then turn most of them off, and add more? Wouldn’t it be clearer to just add the things you want?

Yep, that would be great. `remark-preset-wooorm` has been enabled for a long time, so I don't know are you willing to remove it. Or let's simply use a `prettier` compatible config like https://github.com/1stG/configs/blob/master/packages/remark-config/index.js

Comment (wooorm) [reply docs/404.mdx]: Feel free to add the config you want.

The PR in its current state is broken though, and adding `remark-preset-wooorm` [caught lots of problems](https://github.com/mdx-js/mdx/pull/314/files) that https://github.com/1stG/configs/blob/master/packages/remark-config/index.js won’t catch.

Comment (JounQin) [reply docs/404.mdx]: I'll try to use `.remarkrc.js` instead to resolve the conflict problem first.

Comment (JounQin) [reply docs/404.mdx]: @wooorm OK, I found out the key problem is `remark-retext`. I can not disable `retext-sentence-spacing` directly, but override whole config of `remark-retext` and it works.

https://github.com/mdx-js/mdx/blob/chore/prettier/.remarkrc.js#L10-L18

File: .eslintignore [hunk 2]
```diff
@@ -6,6 +6,7 @@ public
 .next
 .*cache
 *.min.js
+!/.*.js
```
Comment (ChristianMurphy) [inline .eslintignore]: is this intended to ignore many files?
Or just `.remarkrc.js`?

Comment (JounQin) [reply .eslintignore]: .*.js files are ignored by ESLint by default, `!` makes they are not ignored.

File: docs/advanced/transform-content.mdx [hunk 3]
```diff
@@ -19,7 +19,7 @@ Say we have the following file, `example.mdx`:
 import Component from 'Component';
 
 export const meta = {
-    prop: 'value'
+prop: 'value'
```
Comment (wooorm) [inline docs/advanced/transform-content.mdx]: What’s happening here? 🤔

Comment (JounQin) [reply docs/advanced/transform-content.mdx]: It should be `mdx` but the original type is `md`.

Comment (JounQin) [reply docs/advanced/transform-content.mdx]: @wooorm fixed in https://github.com/mdx-js/mdx/pull/927/commits/a5a4fe4bb5973dd24d790a74604c7849e0924f0b

**Comment by JounQin** [comment]
@johno Could you help to uninstall Circle CI? It is really noised after removing its configuration.

File: docs/advanced/ast.mdx [hunk 4]
```diff
@@ -1,11 +1,16 @@
-import { Button } from '@rebass/emotion'
+import {Button} from '@rebass/emotion'
```
Comment (JounQin) [inline docs/advanced/ast.mdx]: Actually I prefer the previous one personally which seems more readable.
Shall/Should we change `bracketSpacing` config of `prettier`?

@ChristianMurphy @wooorm @johno

Comment (JounQin) [reply docs/advanced/ast.mdx]: And about `trailingComma`, `all` is much better for diffing between commits, or `es5` for compatibility.

Comment (wooorm) [reply docs/advanced/ast.mdx]: I personally prefer the “current” style for both (as in: no spaces in braces and no trailing commas)

Comment (JounQin) [reply docs/advanced/ast.mdx]: That's OK, we can discuss it in the future of not, let's just keep it as is in this PR.😀

**Comment by silvenon** [comment]
I'm very late to this. This is generally awesome, I'm just sad about curly braces losing surrounding space:

```diff
-import { Button } from '@rebass/emotion'
+import {Button} from '@rebass/emotion'
```

There's no way around that? (No pun intended.) Prettier adds the space in JS, so I'm confused why it doesn't here.

**Comment by ChristianMurphy** [comment]
It is configurable with https://prettier.io/docs/en/options.html#bracket-spacing

**Comment by silvenon** [comment]
Oh, I didn't realize that we have it explicitly configured to `false`, thanks!

**Pull request merged**